File: sdk-python/src/wzrdmail/_http.py
from __future__ import annotations

import math
import time
from collections.abc import Callable
from email.utils import parsedate_to_datetime
from typing import Any
from urllib.parse import quote

import httpx

from .errors import WzrdmailError

DEFAULT_BASE_URL = "https://api.wzrd.tech"
MAX_RETRIES = 3


def encode_path(segment: str) -> str:
    return quote(segment, safe="")


def _parse_retry_after_ms(header: str | None) -> float | None:
    if header is None or header.strip() == "":
        return None
    try:
        seconds = float(header)
    except ValueError:
        pass
    else:
        if not math.isfinite(seconds):
            return None
        return max(0.0, seconds * 1000)
    try:
        dt = parsedate_to_datetime(header)
    except (TypeError, ValueError):
        return None
    return max(0.0, dt.timestamp() * 1000 - time.time() * 1000)


class HttpClient:
    def __init__(
        self,
        api_key: str | None = None,
        base_url: str | None = None,
        transport: httpx.BaseTransport | None = None,
        timeout: float = 30.0,
    ) -> None:
        self._api_key = api_key
        self._base_url = (base_url or DEFAULT_BASE_URL).rstrip("/")
        self._client = httpx.Client(transport=transport, timeout=timeout)
        self._sleep: Callable[[float], None] = time.sleep

    def close(self) -> None:
        self._client.close()

    def request(
        self,
        method: str,
        path: str,
        *,
        query: dict[str, Any] | None = None,
        body: dict[str, Any] | None = None,
        auth: bool = True,
    ) -> Any:
        headers: dict[str, str] = {"Accept": "application/json"}
        if auth:
            if not self._api_key:
                raise WzrdmailError(
                    0,
                    "unauthorized",
                    "wzrdmail: api_key is required for this call "
                    "(set WZRDMAIL_API_KEY or pass api_key=...)",
                )
            headers["Authorization"] = f"Bearer {self._api_key}"

        params = (
            {k: v for k, v in query.items() if v is not None} if query else None
        )

        attempt = 0
        while True:
            response = self._client.request(
                method,
                f"{self._base_url}{path}",
                params=params,
                json=body,
                headers=headers,
            )
            if response.status_code == 429 and attempt < MAX_RETRIES:
                retry_after_ms = _parse_retry_after_ms(response.headers.get("Retry-After"))
                delay_ms = retry_after_ms if retry_after_ms is not None else (2**attempt) * 500
                self._sleep(delay_ms / 1000)
                attempt += 1
                continue
            break

        if response.status_code >= 400:
            try:
                data = response.json()
            except ValueError:
                data = None
            if isinstance(data, dict) and "name" in data and "message" in data:
                raise WzrdmailError(
                    response.status_code, str(data["name"]), str(data["message"])
                )
            raise WzrdmailError(
                response.status_code,
                "internal_error",
                f"unexpected error response (HTTP {response.status_code})",
            )

        if response.status_code == 204 or not response.content:
            return None
        return response.json()

File: packages/sdk-ts/src/http.ts
import { parseErrorBody } from "./error.js";

export type FetchLike = (input: string, init: RequestInit) => Promise<Response>;

export interface HttpClientOptions {
  apiKey?: string;
  baseUrl?: string;
  fetch?: FetchLike;
  /** Max retries on 429 (expo backoff honoring Retry-After). Default 3. */
  maxRetries?: number;
}

export const DEFAULT_BASE_URL = "https://api.wzrd.tech";

export interface RequestOptions {
  method: "GET" | "POST" | "PATCH" | "DELETE";
  path: string;
  query?: Record<string, string | number | string[] | undefined>;
  body?: unknown;
  auth?: boolean;
}

const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/** Parse Retry-After as delay-seconds or HTTP-date; undefined when absent/malformed. */
const parseRetryAfterMs = (header: string | null): number | undefined => {
  if (header === null || header.trim() === "") return undefined;
  const seconds = Number(header);
  if (Number.isFinite(seconds)) return Math.max(0, seconds * 1000);
  const date = Date.parse(header);
  if (!Number.isNaN(date)) return Math.max(0, date - Date.now());
  return undefined;
};

export class HttpClient {
  private readonly apiKey?: string;
  private readonly baseUrl: string;
  private readonly fetchImpl: FetchLike;
  private readonly maxRetries: number;

  constructor(options: HttpClientOptions) {
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl ?? DEFAULT_BASE_URL).replace(/\/+$/, "");
    this.fetchImpl =
      options.fetch ?? ((input, init) => globalThis.fetch(input, init));
    this.maxRetries = options.maxRetries ?? 3;
  }

  async request<T>(options: RequestOptions): Promise<T> {
    const url = this.buildUrl(options.path, options.query);
    const headers: Record<string, string> = {};
    if (options.auth !== false) {
      if (!this.apiKey) {
        throw new Error(
          "wzrdmail: apiKey is required for this call (set WZRDMAIL_API_KEY or pass { apiKey })"
        );
      }
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    const init: RequestInit = { method: options.method, headers };
    if (options.body !== undefined) {
      headers["Content-Type"] = "application/json";
      init.body = JSON.stringify(options.body);
    }

    let attempt = 0;
    for (;;) {
      const response = await this.fetchImpl(url, init);
      if (response.status === 429 && attempt < this.maxRetries) {
        const retryAfterMs = parseRetryAfterMs(response.headers.get("Retry-After"));
        await sleep(retryAfterMs ?? 2 ** attempt * 500);
        attempt += 1;
        continue;
      }
      if (!response.ok) {
        let raw: unknown;
        try {
          raw = await response.json();
        } catch {
          raw = undefined;
        }
        throw parseErrorBody(response.status, raw);
      }
      if (response.status === 204) return undefined as T;
      return (await response.json()) as T;
    }
  }

  private buildUrl(
    path: string,
    query?: Record<string, string | number | string[] | undefined>
  ): string {
    const url = new URL(`${this.baseUrl}${path}`);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value === undefined) continue;
        url.searchParams.set(
          key,
          Array.isArray(value) ? value.join(",") : String(value)
        );
      }
    }
    return url.toString();
  }
}

export const encodePath = (segment: string): string =>
  encodeURIComponent(segment);
